src/Http/Routing/src/EndpointRoutingMiddleware.cs

```diff
@@ -56,68 +56,75 @@ internal sealed class EndpointRoutingMiddleware
 
         public Task Invoke(HttpContext httpContext)
         {
-            var feature = new EndpointSelectorContext();
+            var feature = new EndpointSelectorContext(httpContext);
+
+            // Store the previous endpoint before matching starts, this lets us restore whatever endpoint was
+            // set at the end of matching if nothing matched
+            var previousEndpoint = feature.Endpoint;
+
+            // We need to set the endpoint to null here so that we can observe any mutations
+            // made by the matcher
+            feature.Endpoint = null;
 
             // There's an inherent race condition between waiting for init and accessing the matcher
             // this is OK because once `_matcher` is initialized, it will not be set to null again.
             var matcherTask = InitializeAsync();
             if (!matcherTask.IsCompletedSuccessfully)
             {
-                return AwaitMatcher(this, httpContext, feature, matcherTask);
+                return AwaitMatcher(this, httpContext, feature, previousEndpoint, matcherTask);
             }
 
             var matchTask = matcherTask.Result.MatchAsync(httpContext, feature);
             if (!matchTask.IsCompletedSuccessfully)
             {
-                return AwaitMatch(this, httpContext, feature, matchTask);
+                return AwaitMatch(this, httpContext, feature, previousEndpoint, matchTask);
             }
 
-            return SetRoutingAndContinue(httpContext, feature);
+            return SetRoutingAndContinue(httpContext, feature, previousEndpoint);
 
             // Awaited fallbacks for when the Tasks do not synchronously complete
-            static async Task AwaitMatcher(EndpointRoutingMiddleware middleware, HttpContext httpContext, EndpointSelectorContext feature, Task<Matcher> matcherTask)
+            static async Task AwaitMatcher(EndpointRoutingMiddleware middleware,
+                                           HttpContext httpContext,
+                                           EndpointSelectorContext feature,
+                                           Endpoint previousEndpoint,
+                                           Task<Matcher> matcherTask)
```

Comment:
aligning parameters is awful, just let vs format it the default way

Reply:
lol sure I prefer alignment but whatevs